### ISSUE TITLE
Firefox: Voice call works only once

### ISSUE THREAD

**Issue by avoidr** [vector-im/riot-web#9031]
### Description

(Voice) call works only (at most) once per Firefox session.
When I try another call, Riot tells me it failed because of permission/mic problem.
In the case when I call and hangup myself (before the call got accepted), the red mic icon (in the URL bar) keeps flashing, as if the microphone was still in use and still recording. Otherwise, the mic gets "freed" correctly and it isn't red and flashing.
Firefox has to be restarted completely in order to successfully call or accept a call. It is not enough to reload the tab.

Browser's permission to access the microphone is granted permanently.
Microphone works in all other programs and there is no actual problem.

I have a friend with the same FF version (65.0.1) and it works perfectly for him.

The same issue does not occur in Chromium (71.0.3578.98).

I am not sure if this is relevant, but I am using the musl libc.

### Steps to reproduce

- Start a voice call
- Cancel call before it is accepted
- Observe, that microphone icon is still active
- Try initiating another call, which fails.

Log: sent    

### Version information

- **Browser**: Mozilla Firefox 65.0.1
- **OS**: VoidLinux with musl libc
- **URL**: riot.im/app

**Comment by avoidr** [comment]
>In the case when I call and hangup myself (before the call got accepted), the red mic icon (in the URL bar) keeps flashing, as if the microphone was still in use and still recording.

It seems that this is not a reliable occurence. But I know it happened a few times before.

**Comment by avoidr** [comment]
To clarify, the calls work only once per Firefox session regardless of whether a call was successfully accepted or not, regardless of the direction of the call.

**Comment by avoidr** [comment]
On start of Firefox (before any calls) Voice & Video settings let me pick a microphone device from a dropdown list. After a call it says no microphones detected.

**Comment by avoidr** [comment]
I see this got a high priority. I think this might actually be a firefox bug in my compilation. But I don't yet know what. I tested mic on http://meet.jit.si/ and my microphone isn't recognised there, too.

My microphone is nothing fancy, has a regular mic jack.

**Comment by jryans** [comment]
Are you able to test a different WebRTC site, such as https://appr.tc/ (note: hosted by Google)?

What behaviour do you see on those other sites?

**Comment by avoidr** [comment]
Same behaviour on https://appr.tc/ too. Microphone permission is requested once, then next times it's as if it didn't exist. Cam works every time.

I guess we're done here, it must be something specific to my setup. I'll look further.

**Comment by jryans** [comment]
I am not sure what to suggest myself. It sounds like potentially a bug in Firefox's device detection if it works on Chrome for your setup.

Perhaps [filing a WebRTC bug](https://bugzilla.mozilla.org/enter_bug.cgi?product=Core&component=WebRTC%3A%20Audio%2FVideo) with Firefox is a good next step.

**Comment by avoidr** [comment]
I have found the solution.

The problem is that firefox stopped supporting ALSA some versions back, in favor of PulseAudio, which I don't have installed. I only have ALSA. Eventually I found a workaround for this:
`apulse firefox`
Literally this is all it took, and I didn't even have to whitelist paths in firefox' sandbox.

Other lessons learned:

- The right mixers have to be enabled in alsamixer, namely "Input Source" had to be set correctly, which I missed at first. Add to this that when I was testing with audacity, audacity actually changes the input source in alsamixer, which explains the "random" behaviour I saw in chromium.
This command helps finding the right configuration¹:
`arecord -vv --format=dat /dev/null`

- Default devices can be set by prioritising kernel modules². Example:
```
$ cat /etc/modprobe.d/alsa-base.conf
options snd_hda_intel index=0
options snd_usb_audio index=1
```
In my case, my webcam was setup as default device if it was plugged in at boot.

¹ https://wiki.archlinux.org/index.php/Advanced_Linux_Sound_Architecture/Troubleshooting#Microphone
² https://wiki.archlinux.org/index.php/Advanced_Linux_Sound_Architecture#Set_the_default_sound_card
